src/gui/csrf.go

```diff
@@ -0,0 +1,117 @@
+package gui
+
+import (
+	"encoding/base64"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/skycoin/skycoin/src/cipher"
+	wh "github.com/skycoin/skycoin/src/util/http" //http,json helpers
+)
+
+const (
+	// the name of CSRF header
+	CSRFHeaderName = "X-CSRF-Token"
+
+	// Max-Age in seconds for cookie. 30 seconds default
+	CSRFMaxAge = time.Duration(30) * time.Second
+
+	// csrf token length
+	csrfTokenLength = 64
+)
+
+type CSRFToken struct {
+	Value     []byte
+	ExpiresAt time.Time
+}
+
+type CSRFStore struct {
+	token   *CSRFToken
+	Enabled bool
+	sync.RWMutex
+}
+
+// getTokenValue returns a url safe base64 encoded token
+func (c *CSRFStore) getTokenValue() string {
+	c.RLock()
+	defer c.RUnlock()
+	return base64.RawURLEncoding.EncodeToString(c.token.Value)
+}
+
+// setToken sets a new CSRF token
+// if the value is changing the expire time should also change
+// so there is no explicit method to just set the value of the token
+func (c *CSRFStore) setToken(token *CSRFToken) {
+	c.Lock()
+	defer c.Unlock()
+	c.token = token
```

Comment:
It's not safe to assign pointer here, the `token` argument can be modified outside of the function.